### PR TITLE
chore(v2): webpack-serve -> webpack-dev-server

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

Switch development server underlying library from `webpack-serve` to `webpack-dev-server`.

Why do we choose `webpack-serve` previously ?

Few months ago, `webpack-dev-server` was in maintenance mode, and then it recommend people to use `webpack-serve` which are said to be more "modern"

However, just recently now `webpack-serve` is DEPRECATED. And suddenly `webpack-dev-server` is in support and update mode again)

There is an extra advantage of this PR: [webpack-dev-server](https://github.com/webpack/webpack-dev-server/) is based on [express](https://github.com/expressjs/express) which is similar to Docusaurus v1 while [deprecated webpack-serve](https://github.com/webpack-contrib/webpack-serve#webpack-serve) is based on  [koa](https://github.com/koajs/koa)

References that I used to migrate:
- https://webpack.js.org/configuration/dev-server/
- https://webpack.js.org/guides/hot-module-replacement/#via-the-nodejs-api
- https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/scripts/start.js

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes
## Test Plan

Locally hot reload still works and nothing wrong with dev server. Performance is similar (not sure which one is better)

### CODE REVIEW

File: packages/docusaurus/lib/webpack/base.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ module.exports = function createBaseConfig(props, isServer) {
     .mode(isProd ? 'production' : 'development')
     .output.path(outDir)
     .filename(isProd ? '[name].[chunkhash].js' : '[name].js')
-    .publicPath(isProd ? baseUrl : '/');
+    .publicPath(baseUrl);
```
Comment (endiliey) [inline packages/docusaurus/lib/webpack/base.js]: This was a bug that I fixed while testing the new PR

File: packages/docusaurus/lib/commands/start.js [hunk 2]
```diff
@@ -97,43 +98,40 @@ module.exports = async function start(siteDir, cliOptions = {}) {
...
+      ignored: /node_modules/,
     },
-  );
+    historyApiFallback: {
+      rewrites: [{from: /\.html$/, to: '/'}],
+    },
+    disableHostCheck: true,
+    overlay: false,
+    host,
+    contentBase: path.resolve(siteDir, 'static'),
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/start.js]: We could potentially add new hooks on this to the ability to execute Express.js middleware

https://webpack.js.org/configuration/dev-server/#devserverafter
https://webpack.js.org/configuration/dev-server/#devserverbefore

```js
before(app, server) {

},
after(app, server) {

}
```

File: packages/docusaurus/lib/commands/start.js [hunk 3]
```diff
@@ -69,15 +67,18 @@ module.exports = async function start(siteDir, cliOptions = {}) {
...
   const host = getHost(cliOptions.host);
   const {baseUrl} = props;
+  const urls = prepareUrls(protocol, host, port);
+  const openUrl = normalizeUrl([urls.localUrlForBrowser, baseUrl]);
 
   // Create compiler from generated webpack config.
   let config = createClientConfig(props);
 
   const {siteConfig, plugins = []} = props;
+  config.plugin('hmr').use(HotModuleReplacementPlugin);
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/start.js]: Needed for hot reload

File: packages/docusaurus/lib/commands/start.js [hunk 4]
```diff
@@ -97,43 +98,40 @@ module.exports = async function start(siteDir, cliOptions = {}) {
...
-        );
-      },
+  // https://webpack.js.org/configuration/dev-server
+  const devServerConfig = {
+    compress: true,
+    clientLogLevel: 'error',
+    hot: true,
+    quiet: true,
+    headers: {
+      'access-control-allow-origin': '*',
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/start.js]: This option was set to enable CQRS.

File: packages/docusaurus/lib/commands/start.js (hunk 3 above)
Comment (yangshun) [reply packages/docusaurus/lib/commands/start.js]: Better to add this as a comment into the code directly. I'll add it.

File: packages/docusaurus/lib/commands/start.js (hunk 4 above)
Comment (yangshun) [reply packages/docusaurus/lib/commands/start.js]: I think you meant CORS. [CQRS](https://martinfowler.com/bliki/CQRS.html) is a different concept altogether.

File: packages/docusaurus/lib/commands/start.js (hunk 2 above)
Comment (yangshun) [reply packages/docusaurus/lib/commands/start.js]: Sounds good! But let's add it when there's a request/we have a need. Gatsby's plugin lifecycles list is so long it's scary.

**Pull request merged**